### DISCUSSION TITLE
Mold as a Library?

### DISCUSSION THREAD

**Discussion by coffeebe4code** [rui314/mold#1218] [General]
Is it possible to use Mold as a library? Is there a particular reason why it cannot be packaged as so?

**Comment by rui314** [comment]
What would be the use case? You can always invoke mold as an external program from any program.

**Reply by coffeebe4code** [reply]
I am making my own programming language, and would prefer to invoke the linker from within my own executable. The obj "files" might already be in memory from previous compiler stages as well. It could reduce code size of the linker itself if certain sections never get called. For example, file loading from previous point. I am no expert in linkers, and I'm not saying I would do what I'm about to say, but other options might not include cross compiler configuration, so all other targets could be removed from the final binary size of my compiler.

also, cute cat!

**Reply by rui314** [reply]
"Linker as a library" is brought up occasionally, but in my opinion, all requests were too vague and are not something that we can implement. ELF and command line options provide a clear user/API interface for the linker, so I'd recommend you stick with them until you find them as a limiting factor in what you are trying to do.

**Reply by heavyrain266** [reply]
"linker as a library" could be useful when combined with `libclang` or `rust-dev` in the buildfarm controllers. Alternatively, for calling the CLI program, I'd love to see the `-json` that will transform output into structured and detailed JSON.

**Reply by rui314** [reply]
Linker's output is binary, I don't know how to want to convert it into JSON.

**Reply by Sir-Walrus** [reply]
Error messages are text

**Reply by rui314** [reply]
Why do you want to see it in json? For IDEs?

**Reply by rui314** [reply]
Anyway, "linker as a library" was the topic of a (too much) heated discussion when I tried to start writing the current version of lld, and I don't want to repeat that. If you want to bring it up again, please show concrete use cases you want to try to implement and convincing argument why the features are not implementable not as a command.

**Reply by heavyrain266** [reply]
I meant generic info and error reports, anything that goes to stdout. If you're calling the `mold` binary as external program that could simplify error reports when you're building the AOSP, other kernels, firmware or games across multiple nodes of the buildfarm. JSON output for errors and any related data could replace the need for library, other tools tends to do that as well (like pijul vcs)

**Reply by rui314** [reply]
That probably should be brought to the compiler first, and the linker would follow suit.
